### PR TITLE
Add support for Hive ACID direct insert

### PR DESCRIPTION
**Pull request by wendigo**
This PR adds preliminary support for reading tables where data was inserted using ACID direct insert (https://issues.apache.org/jira/browse/HIVE-21164)

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -112,16 +112,15 @@
 public class BackgroundHiveSplitLoader
         implements HiveSplitLoader
 {
-    private static final Pattern DIRECT_INSERT_BUCKET_PATTERN = Pattern.compile("bucket_(\\d+)(_\\d+)");
+    private static final Pattern BUCKET_PATTERN = Pattern.compile("bucket_(\\d+)(_\\d+)?$");
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:115]: Not exactly. This pattern will catch both non-acid (6 digits) and acid (5 digits) with optional attemptId.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -534,12 +533,20 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
     private Iterator<InternalHiveSplit> createInternalHiveSplitIterator(Path path, FileSystem fileSystem, InternalHiveSplitFactory splitFactory, boolean splittable, Optional<AcidInfo> acidInfo)
     {
         return Streams.stream(new HiveFileIterator(table, path, fileSystem, directoryLister, namenodeStats, recursiveDirWalkerEnabled ? RECURSE : IGNORED, ignoreAbsentPartitions))
-                .map(status -> splitFactory.createInternalHiveSplit(status, OptionalInt.empty(), splittable, acidInfo))
+                .map(status -> splitFactory.createInternalHiveSplit(status, OptionalInt.empty(), splittable, setIfDirectInsert(acidInfo, status.getPath())))
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:536]: Imho if I'm reading code correctly in loadPartition I'm iterating on directories and here I'm iterating on files (HiveFileIterator) and I'm checking whether bucket file has attemptId

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -648,6 +655,18 @@ static OptionalInt getBucketNumber(String name)
         return OptionalInt.empty();
     }
 
+    static Optional<String> getAttemptId(String bucketFilename)
+    {
+        Matcher matcher = BUCKET_PATTERN.matcher(bucketFilename);
+
+        if (matcher.matches() && matcher.groupCount() == 2) {
+            // https://github.com/apache/hive/blob/ffee30e6267e85f00a22767262192abb9681cfb7/ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java#L444
+            return Optional.ofNullable(matcher.group(2) != null ? matcher.group(2).substring(1) : null);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:664]: I've extracted logic from Hive source code but I'm fine with rewritting it.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java [hunk 4]
```diff
@@ -214,12 +217,20 @@ private void loadValidPositions()
...
     {
-        Path directory = new Path(partitionLocation, deleteDeltaSubdir(
+        Path directory = new Path(acidInfo.getPartitionLocation(), deleteDeltaSubdir(
                 deleteDeltaInfo.getMinWriteId(),
                 deleteDeltaInfo.getMaxWriteId(),
                 deleteDeltaInfo.getStatementId()));
+
+        if (acidInfo.isDirectInsert()) {
+            checkArgument(acidInfo.getBucketId().isPresent(), "bucketId is missing");
+            // When direct insert is enabled (it is by default) base and delta directories contain bucket_[id]_[attemptId] files
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:229]: It defaults to true in Hive sources.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:115]: The problem with this name is that it is not specific. How different is this pattern extracted a `BUCKET_PATTERN` from the other two in `BUCKET_PATTERNS` list?

[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:115]: It's used futher in code to extract attemptId (if it exists) from bucket filename

[the remark this answers is not in the document]
Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:115]: Yeah - I know that. I am just complaining about the name. Not the fact that it is extracted as a constant. Maybe `BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN` then?

[the remark this answers is not in the document]
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:115]: add link to hive code that composes those names

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:537]: And what about `bucketId` parameter passed to `createInternalHiveSplit`. Shouldn't it correspond to bucket id which was possibly discovered in `setIfDirectInsert`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java [hunk 5]
```diff
@@ -214,12 +217,20 @@ private void loadValidPositions()
...
+        Path directory = new Path(acidInfo.getPartitionLocation(), deleteDeltaSubdir(
                 deleteDeltaInfo.getMinWriteId(),
                 deleteDeltaInfo.getMaxWriteId(),
                 deleteDeltaInfo.getStatementId()));
+
+        if (acidInfo.isDirectInsert()) {
+            checkArgument(acidInfo.getBucketId().isPresent(), "bucketId is missing");
+            // When direct insert is enabled base and delta directories contain bucket_[id]_[attemptId] files
+            // but delete delta directories contain bucket files without attemptId.
+            return new Path(directory, format("bucket_%05d", acidInfo.getBucketId().getAsInt()));
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:231]: It should be just enough to cut the last `_.*` part of the original filename, right? It seems a bit safer as we are not recreating whole filename manually which may be error prone. E.g. here we are asssuming that bucket id is formated on 5 digits. while `BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN` does not enforce that.

Also if we go with this approach we could get rid of bucketId from AcidInfo.

Would that work?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 6]
```diff
@@ -625,7 +636,7 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
                 for (LocatedFileStatus file : bucketFiles.get(partitionBucketNumber)) {
                     // OrcDeletedRows will load only delete delta files matching current bucket (same file name),
                     // so we can pass all delete delta locations here, without filtering.
-                    splitFactory.createInternalHiveSplit(file, OptionalInt.of(readBucketNumber), splittable, acidInfo)
+                    splitFactory.createInternalHiveSplit(file, OptionalInt.of(readBucketNumber), splittable, setIfDirectInsert(acidInfo, file.getPath()))
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:639]: the comment above is not valid any more. The filename may be different. Please update.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java (hunk 5 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:231]: Yep, that should work but if hive has direct insert then buckets are always 5 digits long. I can change that to strip attempt id from the string.

Comment (losipiuk) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:231]: I think I would prefer that approach. This way we may have less information in AcidUtils. If we get rid of bucketId from there it makes it easier to reason what is source of truth.

@findepi any though on this one?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:537]: bucketId was removed from AcidInfo

**Comment by losipiuk** [comment]
CI failed: 
```
2020-06-18T13:54:06.3430853Z tests               | 2020-06-18 19:39:06 INFO: FAILURE     /    io.prestosql.tests.hive.TestHiveTransactionalTable.testReadFullAcid (Groups: storage_formats, hive_transactional) took 16.1 seconds
2020-06-18T13:54:06.3455483Z tests               | 2020-06-18 19:39:06 SEVERE: Failure cause:
2020-06-18T13:54:06.3456208Z tests               | java.lang.AssertionError: tableCompactionsIds are empty expected [false] but found [true]
2020-06-18T13:54:06.3456661Z tests               | 	at org.testng.Assert.fail(Assert.java:94)
2020-06-18T13:54:06.3457093Z tests               | 	at org.testng.Assert.failNotEquals(Assert.java:513)
2020-06-18T13:54:06.3457432Z tests               | 	at org.testng.Assert.assertFalse(Assert.java:63)
2020-06-18T13:54:06.3457973Z tests               | 	at io.prestosql.tests.hive.TestHiveTransactionalTable.compactTableAndWait(TestHiveTransactionalTable.java:203)
2020-06-18T13:54:06.3458431Z tests               | 	at io.prestosql.tests.hive.TestHiveTransactionalTable.testReadFullAcid(TestHiveTransactionalTable.java:82)
2020-06-18T13:54:06.3458881Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-06-18T13:54:06.3459245Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-06-18T13:54:06.3459754Z tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-06-18T13:54:06.3460103Z tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-06-18T13:54:06.3460610Z tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-06-18T13:54:06.3461260Z tests               | 	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
2020-06-18T13:54:06.3461727Z tests               | 	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
2020-06-18T13:54:06.3462089Z tests               | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2020-06-18T13:54:06.3463249Z tests               | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2020-06-18T13:54:06.3463683Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-06-18T13:54:06.3463990Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-06-18T13:54:06.3464427Z tests               | 	at java.base/java.lang.Thread.run(Thread.java:834)
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 7]
```diff
@@ -531,12 +534,20 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
...
-                .map(status -> splitFactory.createInternalHiveSplit(status, OptionalInt.empty(), splittable, acidInfo))
+                .map(status -> splitFactory.createInternalHiveSplit(status, OptionalInt.empty(), splittable, setIfDirectInsert(acidInfo, status.getPath())))
                 .filter(Optional::isPresent)
                 .map(Optional::get)
                 .iterator();
     }
 
+    private static Optional<AcidInfo> setIfDirectInsert(Optional<AcidInfo> acidInfo, Path filename)
+    {
+        Optional<String> attemptId = getAttemptId(filename.getName());
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:545]: nit: you  can put it inside lambda

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java [hunk 8]
```diff
@@ -214,12 +216,20 @@ private void loadValidPositions()
...
+    private static Path createPath(AcidInfo acidInfo, AcidInfo.DeleteDeltaInfo deleteDeltaInfo, String fileName)
     {
-        Path directory = new Path(partitionLocation, deleteDeltaSubdir(
+        Path directory = new Path(acidInfo.getPartitionLocation(), deleteDeltaSubdir(
                 deleteDeltaInfo.getMinWriteId(),
                 deleteDeltaInfo.getMaxWriteId(),
                 deleteDeltaInfo.getStatementId()));
+
+        if (acidInfo.isDirectInsert()) {
+            verify(CharMatcher.is('_').countIn(fileName) >= 2, "Filename should contain attemptId");
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:227]: going further from this place it look that actually we do not need even `directId` flag in `AcidInfo`. Right? We can just check for attemptId in fileName here.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java (hunk 5 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java]: i think hive's logic is to recreate the filename. While it seems more err-prone, it may also be more future proof.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java [hunk 9]
```diff
@@ -161,22 +172,34 @@ public static Builder builder(Path partitionPath)
...
         private Builder(Path partitionPath)
         {
-            partitionLocation = requireNonNull(partitionPath, "partitionPath is null");
+            partitionLocation = requireNonNull(partitionPath, "partitionPath is null").toString();
+        }
+
+        private Builder(AcidInfo acidInfo)
+        {
+            partitionLocation = acidInfo.getPartitionLocation();
+            deleteDeltaInfoBuilder.addAll(acidInfo.deleteDeltas);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java:194]: Copy directInsert field too.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java [hunk 10]
```diff
@@ -161,22 +172,34 @@ public static Builder builder(Path partitionPath)
...
 
+    public static Builder builder(AcidInfo acidInfo)
+    {
+        return new Builder(acidInfo);
+    }
+
     public static class Builder
     {
-        private final Path partitionLocation;
+        private final String partitionLocation;
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java:182]: This change seems unrelated. Consider reverting.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 11]
```diff
@@ -645,6 +656,17 @@ static OptionalInt getBucketNumber(String name)
         return OptionalInt.empty();
     }
 
+    @VisibleForTesting
+    static Optional<String> getAttemptId(String bucketFilename)
+    {
+        Matcher matcher = BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN.matcher(bucketFilename);
+        if (matcher.matches() && matcher.groupCount() == 2 && matcher.group(2) != null) {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:663]: checking `matcher.groupCount() == 2` is wrong, since the group count is statically defined in the pattern

```suggestion
        if (matcher.matches() && matcher.group(2) != null) {
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java (hunk 8 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:227]: bump

**Comment by findepi** [comment]
CI failed -- https://github.com/prestosql/presto/issues/4173

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java (hunk 8 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:227]: I'm not really sure if it's safe to assume that if filename has 2 x _ then second substring is an attemptId

Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java:227]: I agree, but this is how we recognize them. More precisely, we recognize them by BackgroundHiveSplitLoader.BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN (so: 2 x underscore, literal "bucket" and some digits). It doesn't hurt any more to use this logic here as well.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java [hunk 12]
```diff
@@ -645,6 +656,17 @@ static OptionalInt getBucketNumber(String name)
         return OptionalInt.empty();
     }
 
+    @VisibleForTesting
+    static Optional<String> getAttemptId(String bucketFilename)
+    {
+        Matcher matcher = BUCKET_WITH_OPTIONAL_ATTEMPT_ID_PATTERN.matcher(bucketFilename);
+        if (matcher.matches() && matcher.group(2) != null) {
+            return Optional.of(matcher.group(2).substring(1));
+        }
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java:665]: since we generally discard attemptId information, this could be `hasAttemptId`
